### PR TITLE
Support N-dimensional deconvolution function and link.

### PR DESCRIPTION
**Pull request by takagi**
This PR generalizes two-dimensional deconvolution function and link to N-dimensional ones.

### CODE REVIEW

**Comment by takagi** [comment]
I show current status of the first implementation of this PR.

■…Done □…Doing

**ND Deconvolution**

|Items|Documentation|Test|Implementation|
|---|---|---|---|
|Function - CPU - forward|N/A|□|□|
|Function - CPU - backward|N/A|□|□|
|Function - GPU - forward|N/A|□|□|
|Function - GPU - backward|N/A|□|□|
|Function - deconvolution_nd|□|□|□|
|Link - DeconvolutionND|□|□|□|

**Comment by takagi** [comment]
I will add tests for FP16 and FP64 after #1295, dtype support for initializers, is merged.

File: chainer/functions/connection/deconvolution_nd.py [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+        dims = x.shape[2:]      # x: n, C_I, d_1, d_2, ..., d_N
+        stride = self.stride
+        pad = self.pad
+
+        # gcol: C_O, k_1, ..., k_N, n, d_1, ..., d_N
+        gcol = xp.tensordot(W, x, (0, 1)).astype(x.dtype)
+        # Roll n, which is batch size, before the first.
+        gcol = xp.rollaxis(gcol, ndim + 1)
+
+        if self.outs is None:
```
Comment (delta2323) [inline chainer/functions/connection/deconvolution_nd.py]: Can we calculate `self.outs` at `__init__`?

File: chainer/functions/connection/deconvolution_nd.py [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+        ndim = self.ndim
+        colon = slice(None)
+
+        # Make empty array for output.
+        if self.outs is None:
+            self.outs = tuple(
+                conv.get_deconv_outsize(d, k, s, p)
+                for d, k, s, p in zip(dims, ksize, self.stride, self.pad))
+        y_shape = (n, c) + self.outs  # (n, c_O, out_1, out_2, ..., out_N)
+        y = cuda.cupy.empty(y_shape,  dtype=x.dtype)
```
Comment (delta2323) [inline chainer/functions/connection/deconvolution_nd.py]: Too many spaces.

File: chainer/links/connection/deconvolution_nd.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+            initializer instance or another value that
+            :func:`~chainer.init_weight` helper function can take. This link
+            uses :func:`~chainer.init_weight` to initialize the filter weight
+            and passes the value of ``initialW`` to it as it is.
+        initial_bias: Value used to initialize the bias vector. May be an
+            initializer instance or another value except ``None`` that
+            :func:`~chainer.init_weight` helper function can take. If ``None``
+            is given, this link does not use the bias vector. This link uses
+            :func:`~chainer.init_weight` to initialize the bias vector and
+            passes the value of ``initial_bias`` other than ``None`` to it as
```
Comment (delta2323) [inline chainer/links/connection/deconvolution_nd.py]: I think "other than None" is not needed because we explained it earlier.

File: tests/chainer_tests/links_tests/connection_tests/test_deconvolution_nd.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        in_channels = 3
+        out_channels = 2
+        ndim = len(self.dims)
+        ksize = (3,) * ndim
+        stride = (2,) * ndim
+        pad = (1,) * ndim
+
+        outs = tuple(
+            conv.get_deconv_outsize(d, k, s, p)
+            for (d, k, s, p) in zip(self.dims, ksize, stride, pad))
```
Comment (delta2323) [inline tests/chainer_tests/links_tests/connection_tests/test_deconvolution_nd.py]: I would like you to check the case in which outsize is determined by user, but its value is different from one determined by `get_deconv_outsize`.

When we fix the output size of convolution, there are several input sizes of the convolution that outputs the specified output size. (i.e. the equation get_conv_outsize(input_shape) == output_shape has multiple solutions of input_shape). So, when we make an autoencoder with a convolution and a deconvolution, we cannot determine the output size of the deconvolution solely from its input. This is why `Deconvolution2D` has `outsize` argument. `get_deconv_outsize` returns just one of the possible output size (maybe smallest or largest). So I would like you to check the behavior of `Deconvolution2D` with the condition mentioned above.

File: chainer/functions/connection/deconvolution_nd.py (hunk 2 above)
Comment (takagi) [inline chainer/functions/connection/deconvolution_nd.py]: oh...

File: chainer/links/connection/deconvolution_nd.py (hunk 3 above)
Comment (takagi) [inline chainer/links/connection/deconvolution_nd.py]: Sure.

File: tests/chainer_tests/links_tests/connection_tests/test_deconvolution_nd.py (hunk 4 above)
Comment (takagi) [inline tests/chainer_tests/links_tests/connection_tests/test_deconvolution_nd.py]: Sure, I will add tests with output size with values `get_deconv_outsize` does not give, thanks for the clear explanation.

Comment (takagi) [inline tests/chainer_tests/links_tests/connection_tests/test_deconvolution_nd.py]: I added the test.

File: chainer/link.py [hunk 5]
```diff
@@ -97,8 +120,9 @@ def __init__(self, **params):
         self._cpu = True
         self.name = None
 
-        for name, shape in six.iteritems(params):
-            self.add_param(name, shape)
+        for name, value in six.iteritems(params):
+            shape, dtype = _ensure_shape_dtype(value)
```
Comment (delta2323) [inline chainer/link.py]: Is this change related to ConvolutionND?

Comment (takagi) [inline chainer/link.py]: It was merged from the branch of https://github.com/pfnet/chainer/pull/1633, it is now merged into master so the diff will go away after updating this PR.

**Comment by takagi** [comment]
I resolved the conflict.

File: chainer/link.py (hunk 5 above)
Comment (takagi) [inline chainer/link.py]: It's gone away.

**Comment by takagi** [comment]
I will fix this PR to use `Link.add_param` 's `initializer` parameter.

File: chainer/functions/connection/deconvolution_nd.py [hunk 6]
```diff
@@ -303,12 +303,12 @@ def deconvolution_nd(x, W, b=None, stride=1, pad=0, outsize=None,
     weight ``W``, and the bias vector ``b``.
 
     Args:
-        x (chainer.Variable or :class:`numpy.ndarray` or cupy.ndarray):
-            Input variable of shape :math:`(n, c_I, d_1, d_2, ..., d_N)`.
-        W (chainer.Variable or :class:`numpy.ndarray` or cupy.ndarray):
-            Weight variable of shape :math:`(c_O, c_I, k_1, k_2, ..., k_N)`.
-        b (chainer.Variable or :class:`numpy.ndarray` or cupy.ndarray):
-            Bias variable of length :math:`c_O` (optional).
+        x (~chainer.Variable or :class:`numpy.ndarray` or cupy.ndarray):
```
Comment (delta2323) [inline chainer/functions/connection/deconvolution_nd.py]: As you can see by compiling the document, `~`  does not work in this case (Napoleon cannot seem to recognize it as a class when the classname is followed by some words).

File: chainer/functions/connection/deconvolution_nd.py [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+            ``(s, s, ..., s)``.
+        pad (int or tuple of ints): Spatial padding size for input arrays
+            :math:`(p_1, p_2, ..., p_N)`. ``pad=p`` is equivalent to
+            ``(p, p, ..., p)``.
+        outsize (tuple of ints): Expected output size of deconvolutional
+            operation. It should be a tuple of ints
+            :math:`(out_1, out_2, ..., out_N)`. Default value is ``None`` and
+            the outsize is estimated by input size, stride and pad.
+        use_cudnn (bool): If ``True``, then this function uses cuDNN if
+            available. Deconvolution using cuDNN supports more than
```
Comment (delta2323) [inline chainer/functions/connection/deconvolution_nd.py]: Note that cuDNN supports more than one-dimensional deconvolution operations only.

**Comment by delta2323** [comment]
Could you add test cases in which `check_type_forward` fail if it is possible?

File: chainer/functions/connection/deconvolution_nd.py (hunk 6 above)
Comment (takagi) [inline chainer/functions/connection/deconvolution_nd.py]: oh...

Comment (takagi) [inline chainer/functions/connection/deconvolution_nd.py]: Fixed!

**Comment by takagi** [comment]
> Could you add test cases in which check_type_forward fail if you have a time?
Sure.

**Comment by takagi** [comment]
I've added tests for type check.

**Comment by delta2323** [comment]
LGTM 👍
